float32 dist_error
float32 theta_error